### DISCUSSION TITLE
gslc_ElemSetTxtStr not updating the text

### DISCUSSION THREAD

**Discussion by PrattTechnologies** [ImpulseAdventure/GUIslice#386] [Q&A] [answered]
Hi
I am trying to update the text of a dynamic text field created in GUISlice Builder. The code displays the GUI but does not update the text element on the GUI, could someone point out what I am missing? Thanks in advance.

Properties of the text from GUISlice Builder:
ElementRef: `pElemDol1`
External Storage Size: `7`

The code from Arduino IDE (snipped):

```
#include "Arduino.h"
#include "test_GSLC.h"

// gui elements
#define MAX_STR 100
gslc_tsElemRef* pElemDol1         = NULL;

// dec functions
void lcdFunc();

void setup() 
{
    // Initiate Serial
    Serial.begin(9600);
    while (!Serial) ; // wait for serial port to connect

    // Initiate the LCD
    gslc_InitDebug(&DebugOut);
    InitGUIslice_gen();
}

void loop() 
{
    gslc_Update(&m_gui);
    lcdFunc();
}

void lcdFunc()
{
  char acTxt[MAX_STR];
  snprintf(acTxt, MAX_STR, "%s", "1234");
  gslc_ElemSetTxtStr(&m_gui, pElemDol1, "acTxt");
}
```

**Comment by Pconti31** [comment, marked as answer]
@PrattTechnologies While acTxt will hold MAX_STR, pElemDol1 will only hold 6 characters plus the null so that's a dangerous way to code.  No point in using snprintf if using MAX_STR as the gate value.  Try,
```
snprintf(acTxt, 7, "%s", "1234");
gslc_ElemSetTxtStr(&m_gui, pElemDol1, acTxt);
```
Note the missing quotes around acTxt.

Now all that said you still should have gotten "acTxt" as a string output on your display.
If not, zip the entire project folder and post it so we can see what is going on.

You also might want to read the Builder's faq to see how to place numbers into a dynamic text field since snprintf won't do that with an Arduino.
Paul--

**Reply by Pconti31** [reply]
Also check out example ex04 and how we handle the counter.
Paul--

**Reply by PrattTechnologies** [reply]
Hi @Pconti31 
Sorry for the late reply, I never got a notification for this.

Note taken regarding the string length. The only way I got this to work was to take it out of it's own function and have the update done in the void loop() section.

Many thanks again.

**Reply by elmosolutions** [reply]
I have been unsuccessful so far in getting the function to work.

Out of despair, I even added a line that just tries to update the string value right after the creation code:

  // Create E_ELEM_TEXT9 text label
  pElemRef = gslc_ElemCreateTxt(&m_gui,E_ELEM_TEXT9,E_PG_Interrupteurs,(gslc_tsRect){417,292,44,17},
    (char*)"99.9",0,E_FREESANSBOLD12);
  gslc_ElemSetTxtCol(&m_gui,pElemRef,GSLC_COL_WHITE);
  gslc_ElemSetCol(&m_gui,pElemRef,GSLC_COL_BLACK,GSLC_COL_BLUE_DK2,GSLC_COL_BLACK);
  **gslc_ElemSetTxtStr(&m_gui, pElemRef, "36.9");**
  
The value stays at "99.9"
  
Still, Example 04 works great on a Mega 2560 board.

Any ideas?

Thanks,

Ricardo.

**Comment by Pconti31** [comment]
@elmosolutions Yes, its still the same mistake.
```
pElemRef = gslc_ElemCreateTxt(&m_gui,E_ELEM_TEXT9,E_PG_Interrupteurs,(gslc_tsRect){417,292,44,17},
(char*)"99.9",0,E_FREESANSBOLD12);
```
You are trying to replace flash storage "99.9" length "0" which is READ-ONLY.  If you are using the Builder then for this field's property values set: 
```
ElementRef=m_pElemTemperature
External Storage Size = 7
```
This should create lines in your *_GSLC.h file something like:
```
  // Create E_ELEM_TEXT9 runtime modifiable text
  static char m_sDisplayText9[8] = "99.9";
  pElemRef = gslc_ElemCreateTxt(&m_gui,E_ELEM_TEXT9,E_PG_Interrupteurs,(gslc_tsRect){417,292,44,17},
    (char*)m_sDisplayText9,8,E_FREESANSBOLD12);
  gslc_ElemSetTxtCol(&m_gui,pElemRef,GSLC_COL_BLACK);
  gslc_ElemSetCol(&m_gui,pElemRef,GSLC_COL_GRAY,GSLC_COL_WHITE,GSLC_COL_WHITE);
  m_pElemTemperature = pElemRef;
```
Note static char  m_sDisplayText9[8] is now created in SRAM writeable memory and GUIslice's gslc_ElemCreateTxt(() storage length is changed from 0 to 8! now in your loop() not in the InitGUIslice_gen() you can use
```
gslc_ElemSetTxtStr(&m_gui, m_pElemTemperature, "36.9");
```
Good luck!
Paul--

**Reply by elmosolutions** [reply]
Paul, that worked like a charm. Thanks a million!

Is a static declaration the only way to be guarantee that the declared object will be read/write?

Ricardo.

**Comment by Pconti31** [comment]
@elmosolutions No.  if a an object is declared const than its in flash, like const int, const char, or in your case (char*)"99.9" etc... 
The static just makes it easier than declaring the object's address in global space.  If the builder had used char m_sDisplayText9[8]
instead of static char m_sDisplayText9[8] then m_sDisplayText9 would have been placed on the system stack and would have been deleted or popped off the stack when the program exited from the InitGUIslice_gen() routine so a crash would occur if GUIslice tried to access the memory even though it was writable since the memory would have been deleted.
Paul--

**Reply by elmosolutions** [reply]
Got it. Thanks again!

Ricardo.
